### ISSUE TITLE
any idea how to use sentence encoder as Siamese architecture?

### ISSUE THREAD

**Issue by omerarshad** [tensorflow/hub#87]
how can we use sentence encoder as siamese architecture to pass two sentences as input and fine tune encoder aswell?

**Comment by andresusanopinto** [comment]
Yes. Modules can be applied more than once, so you can do things like:

```python
m = hub.Module("https://tfhub.dev/google/universal-sentence-encoder/2")
embedding1 = m(query_sentences)
embedding2 = m([example_sentence])
outputs = tf.matmul(query_sentences, example_sentence)
```

**Comment by catalla** [comment]
When re-training the encoder as a Siamese network, are gradients guaranteed to propagate correctly even if the module is run twice shown above?

We cannot specify 'reuse=True' so, it seems unclear whether applying the module a second time will erase the state after the first time.

```python
m = hub.Module("https://tfhub.dev/google/universal-sentence-encoder/2", trainable=True)
embedding1 = m(query_1)
embedding2 = m(query_2)

loss = tf.square(embedding1 - embedding2) # Some loss function here
optimizer = tf.train.GradientDescentOptimizer().minimize(loss)
sess.run(optimizer)
```

**Comment by meethariprasad** [comment]
> Yes. Modules can be applied more than once, so you can do things like:
> 
> ```python
> m = hub.Module("https://tfhub.dev/google/universal-sentence-encoder/2")
> embedding1 = m(query_sentences)
> embedding2 = m([example_sentence])
> outputs = tf.matmul(query_sentences, example_sentence)
> ```

Can you please explain outputs layer here please? how this example works with say dense layer?

**Comment by ARArun** [comment]
@meethariprasad Have you figured out how to use USE as a Siamese Architecture

**Comment by meethariprasad** [comment]
> @meethariprasad Have you figured out how to use USE as a Siamese Architecture

@ARArun: No, Not tried, not because I was not able to do it, because theoratically after thinking through, not found strong evidence of improvement.  That is because I found out that Siamese as in MaLSTM makes sense in Word Embedding to Sentence Construction Scenario. In case of pretrained sentence embedding converting objective in to classification itself achieves the fine tuning of use based on similiar sentences. MaLSTM Siamese looks over complex solution as you need to break it to LSTM sequence of word for that get word embedding from USE (Which is not the right way to use USE as Transformer Network applies positional weights on words based on entire sentence). 

To conflict with myself on above statement, one use case of MaLSTM is thinking sentences itself as sequence of steps. That is interesting scenario, but I don't have any concrete use case to define on it here.

Any difference in opinion will be highly appreciated so that I can learn from it.
